### PR TITLE
feat: `Page.pop_views_until()` to pop multiple views and return a result to the destination view

### LINKED ISSUE #6326: feature: Add `pop_until_with_result` to Page - pop multiple views and return a result
### Duplicate Check

- [x] I have searched the [opened issues](https://github.com/flet-dev/flet/issues) and there are no duplicates

### Describe the requested feature

Currently, Flet has no way to pop multiple views from the navigation stack **and return a result** to the destination view. Developers who need to navigate back through several pages (e.g., from a multi-step flow) and pass data back to the originating view must resort to shared application state or manual workarounds.

Flutter recently added [`Navigator.popUntilWithResult`](https://api.flutter.dev/flutter/widgets/Navigator/popUntilWithResult.html), which solves this exact problem on the imperative side. Flet should expose an equivalent in its declarative navigation model.

**References:**
- [Flutter PR #169341](https://github.com/flutter/flutter/pull/169341)
- [Flutter Issue #30112](https://github.com/flutter/flutter/issues/30112) — 79+ reactions, open for 6 years (2019–2025)
- [API Docs](https://api.flutter.dev/flutter/widgets/Navigator/popUntilWithResult.html)
- [Short Video](https://youtube.com/shorts/Z3JccIFWZFs?si=PJF9DTRm9Jr9GUBL)

---

#### Flutter context

Flutter's [issue #30112](https://github.com/flutter/flutter/issues/30112) was opened in **March 2019** and received **79+ reactions**. It took **6 years** to be resolved (merged October 2025, shipped in Flutter 3.41.0).

The problem: `Navigator.pop()` accepts an optional `result`, but `Navigator.popUntil()` — which pops multiple routes — did **not**. Developers were forced into workarounds:

| Workaround | Problem |
|---|---|
| Chained `.pop()..pop()..pop(true)` | Brittle; requires knowing exact stack depth |
| Provider / Global State | Overkill for simple navigation results |
| `pushNamedAndRemoveUntil` with arguments | Wrong animation direction (push instead of pop) |
| `shared_preferences` | Extremely heavy; race conditions |
| `Completer<T>` passed through stack | Complex plumbing; easy to leak |

Developer quotes from the issue:

> *"Very ugly for something that should've been provided out of the box by Flutter."* — **radiKal07**

> *"I have no clue why is this not implemented yet."* — **topnax**

#### The same problem in Flet

The problem **repeats in Flet**, and is arguably **worse** because Flet's declarative model makes the workarounds more verbose.

The existing `ViewPopEvent` has **no `result` field**:

```python
class ViewPopEvent(Event["Page"]):
    route: str
    view: Optional[View] = None
    # No result!
```

**Workarounds Flet developers must use today:**

**1. Global shared state (most common)**
```python
result_holder = {"value": None}

async def finish_flow(e):
    result_holder["value"] = "completed"  # write to global
    while len(page.views) > 1:
        page.views.pop()
    await page.push_route("/")

def route_change():
    # ... rebuild views ...
    if result_holder["value"]:  # check global manually
        show_result(result_holder["value"])
        result_holder["value"] = None  # manual cleanup
```

**2. Session store**
```python
page.session.store.set("flow_result", "completed")
# ... later ...
result = page.session.store.get("flow_result")
```

**3. Manual view stack loop**
```python
while len(page.views) > 1 and page.views[-1].route != "/":
    page.views.pop()
await page.push_route("/")
# Result? Still need global state.
```

All of these suffer from: no type-safety, manual cleanup, race conditions, tight coupling between views, and excessive boilerplate.

---

### Describe the solution you'd like

A new method on `Page`:

```python
await page.pop_until_with_result(route: str, result: Any = None)
```

Along with a new event:

```python
page.on_view_pop_result = handler  # receives ViewPopResultEvent
```

#### Before vs After

**BEFORE (workaround with global state):**
```python
result_holder = {"value": None}

def main(page: ft.Page):
    def route_change():
        page.views.clear()
        page.views.append(ft.View(route="/", controls=[
            ft.Text(f"Result: {result_holder['value']}"),
            ft.Button("Start", on_click=lambda _:
                asyncio.create_task(page.push_route("/step1"))),
        ]))
        if page.route == "/step2":
            page.views.append(ft.View(route="/step2", controls=[
                ft.Button("Finish", on_click=lambda _: finish()),
            ]))
        if result_holder["value"]:
            page.show_dialog(ft.SnackBar(ft.Text(result_holder["value"])))
            result_holder["value"] = None
        page.update()

    async def finish():
        result_holder["value"] = "Flow completed!"
        while len(page.views) > 1:
            page.views.pop()
        await page.push_route("/")
```

**AFTER (with `pop_until_with_result`):**
```python
def main(page: ft.Page):
    def route_change():
        page.views.clear()
        page.views.append(ft.View(route="/", controls=[...]))
        if page.route == "/step2":
            page.views.append(ft.View(route="/step2", controls=[
                ft.Button("Finish", on_click=lambda _:
                    asyncio.create_task(
                        page.pop_until_with_result("/", result="Flow completed!")
                    )),
            ]))
        page.update()

    def on_pop_result(e: ft.ViewPopResultEvent):
        page.show_dialog(ft.SnackBar(ft.Text(f"Result: {e.result}")))
        page.update()

    page.on_route_change = route_change
    page.on_view_pop_result = on_pop_result
```

| Aspect | Before | After |
|---|---|---|
| Result passing | Global state or closure | `result` in event |
| Data cleanup | Manual | Automatic |
| Type-safety | None | `ViewPopResultEvent.result` |
| View removal | Manual loop | One method call |
| Route update | Manual (`push_route` + `update`) | Automatic |
| Lines of code | ~15-20 for the workaround | 1 call + 1 handler |
| Bug risk | High (race conditions, missing view) | Low (clear `ValueError`) |

---

#### Architecture context

Flet uses a **declarative `Navigator`** (based on `pages: [...]`), not imperative `push`/`pop`. The `page.views` list in Python is the source of truth and is synced to the Dart client's `Navigator`. Because of this, Flutter's `Navigator.popUntilWithResult` cannot be called directly — the equivalent semantics are implemented within Flet's own navigation model.

**No Dart-side changes are needed.** The Python-side `page.update()` syncs the updated view list to the Dart client, and the Flutter `Navigator` rebuilds with correct transition animations automatically.

### Edge cases

| Scenario | Expected behavior |
|----------|-------------------|
| Target route not found in `page.views` | Raise `ValueError` |
| Target route is already the top view | No views popped; only `on_view_pop_result` fires |
| Only one view in stack (and it's the target) | Only `on_view_pop_result` fires |
| Multiple views with the same route | First match from the bottom (like Flutter's `ModalRoute.withName`) |

### Suggest a solution

_No response_

### Screenshots

Image

### Additional details

I already have a working implementation with unit tests and example app ready to submit as a PR.

### PR DESCRIPTION
**Pull request by brunobrown**
## Description

Add `Page.pop_until_with_result()` method and `ViewPopResultEvent` to support popping multiple views from the navigation stack and returning a result to the destination view — the Flet equivalent of Flutter's [`Navigator.popUntilWithResult`](https://api.flutter.dev/flutter/widgets/Navigator/popUntilWithResult.html).

Fixes #6326

## Test Code

```python
import asyncio

import flet as ft


def main(page: ft.Page):
    page.title = "pop_until_with_result Example"

    result_text = ft.Text("No result yet", size=18)

    def route_change():
        page.views.clear()
        page.views.append(
            ft.View(
                route="/",
                controls=[
                    ft.AppBar(title=ft.Text("Home"), bgcolor=ft.Colors.SURFACE_BRIGHT),
                    result_text,
                    ft.Button(
                        "Start flow",
                        on_click=lambda _: asyncio.create_task(page.push_route("/step1")),
                    ),
                ],
            )
        )
        if page.route in ["/step1", "/step2"]:
            page.views.append(
                ft.View(
                    route="/step1",
                    controls=[
                        ft.AppBar(title=ft.Text("Step 1"), bgcolor=ft.Colors.SURFACE_BRIGHT),
                        ft.Button(
                            "Go to Step 2",
                            on_click=lambda _: asyncio.create_task(page.push_route("/step2")),
                        ),
                    ],
                )
            )
        if page.route == "/step2":
            page.views.append(
                ft.View(
                    route="/step2",
                    controls=[
                        ft.AppBar(title=ft.Text("Step 2 (Final)"), bgcolor=ft.Colors.SURFACE_BRIGHT),
                        ft.Text("Flow complete!"),
                        ft.Button(
                            "Finish and go Home",
                            on_click=lambda _: asyncio.create_task(
                                page.pop_until_with_result("/", result="Flow completed!")
                            ),
                        ),
                    ],
                )
            )
        page.update()

    def on_pop_result(e: ft.ViewPopResultEvent):
        result_text.value = f"Result: {e.result}"
        page.show_dialog(ft.SnackBar(ft.Text(f"Result: {e.result}")))
        page.update()

    async def view_pop(e: ft.ViewPopEvent):
        if e.view is not None:
            page.views.remove(e.view)
            top_view = page.views[-1]
            await page.push_route(top_view.route)

    page.on_route_change = route_change
    page.on_view_pop = view_pop
    page.on_view_pop_result = on_pop_result
    route_change()


if __name__ == "__main__":
    ft.run(main)
```

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I signed the CLA.
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally with my changes
- [ ] I have made corresponding changes to the [documentation](https://github.com/flet-dev/website) (if applicable)

## Screenshots

pop-until-with-result-example

## Additional details

- **Flutter reference:** [Navigator.popUntilWithResult](https://api.flutter.dev/flutter/widgets/Navigator/popUntilWithResult.html) — [issue #30112](https://github.com/flutter/flutter/issues/30112) (79+ reactions, 6 years open)
- **No Dart-side changes needed** — Python's `page.update()` syncs the view list and Flutter's Navigator rebuilds with correct transitions automatically
- **12 unit tests** covering: event creation, exports, method behavior, edge cases, event firing
- **Example app** added at `examples/apps/routing_navigation/pop_until_with_result.py`

### Files changed

| File | Change |
|------|--------|
| `sdk/python/packages/flet/src/flet/controls/page.py` | `ViewPopResultEvent`, `pop_until_with_result()`, `on_view_pop_result`, unified `before_event` |
| `sdk/python/packages/flet/src/flet/__init__.py` | Export `ViewPopResultEvent` |
| `sdk/python/packages/flet/tests/test_pop_until_with_result.py` | 12 unit tests |
| `sdk/python/examples/apps/routing_navigation/pop_until_with_result.py` | Example app |
| `sdk/python/packages/flet/docs/assets/navigation-routing/pop-until-with-result-example.gif` | Demo GIF |

## Summary by Sourcery

Add support for popping multiple views with a result and wire it into page navigation events.

New Features:
- Introduce ViewPopResultEvent to carry navigation results back to the destination view when popping views.
- Add Page.pop_until_with_result() to pop back to a target route while returning a result via on_view_pop_result.
- Expose ViewPopResultEvent from the top-level flet package for direct import.
- Provide an example routing app demonstrating pop_until_with_result usage.

Tests:
- Add unit tests covering ViewPopResultEvent creation, typing, imports, the new on_view_pop_result handler, and pop_until_with_result behavior and edge cases.

### CODE REVIEW

**Comment by FeodorFitsner** [comment]
Thanks for the PR!

A few notes:
- Rename `pop_until_with_result()` to `pop_views_until()` (with optional result).
- Rename `on_view_pop_result` to `on_views_pop_until`.

File: sdk/python/packages/flet/tests/test_pop_views_until.py
Comment (ndonkoHenri) [inline sdk/python/packages/flet/tests/test_pop_views_until.py:1]: Instead of a unit test, i will suggest an integration test, similar to https://github.com/flet-dev/flet/blob/main/sdk/python/packages/flet/integration_tests/examples/apps/test_routing_navigation.py.

**Comment by FeodorFitsner** [comment]
Hey @brunobrown, can you change the base of your PR to `release/v0.85.0`, pull the latest from the base branch and fix integration test(s). Thanks!

**Comment by FeodorFitsner** [comment]
Could you resolve conflicts please?

File: sdk/python/packages/flet/src/flet/controls/page.py [hunk 1]
```diff
@@ -506,6 +533,14 @@ class Page(BasePage):
     control.
     """
 
+    on_views_pop_until: Optional[EventHandler[ViewsPopUntilEvent]] = None
+    """
+    Called when [`pop_views_until`][flet.Page.pop_views_until] reaches
+    the destination view.
```
Comment (ndonkoHenri) [inline sdk/python/packages/flet/src/flet/controls/page.py]: Please update your cross references in the modified/added docstrings to use the new style: https://github.com/flet-dev/flet/blob/main/.codex/skills/docs-conventions/SKILL.md?plain=1#L12

File: sdk/python/packages/flet/src/flet/controls/page.py [hunk 2]
```diff
@@ -897,6 +932,75 @@ async def view_pop(e):
             arguments={"route": new_route},
         )
 
+    async def pop_views_until(self, route: str, result: Any = None) -> None:
+        """
+        Pops views from the navigation stack until a view with the given
+        `route` is found, then delivers `result` via the
+        [`on_views_pop_until`][flet.Page.on_views_pop_until] event.
+
+        This is the Flet equivalent of Flutter's `Navigator.popUntilWithResult`.
```
Comment (ndonkoHenri) [inline sdk/python/packages/flet/src/flet/controls/page.py]: Users dont usually have flutter knowledge, so we can remove this line.

File: sdk/python/packages/flet/src/flet/controls/page.py [hunk 3]
```diff
@@ -897,6 +932,75 @@ async def view_pop(e):
             arguments={"route": new_route},
         )
 
+    async def pop_views_until(self, route: str, result: Any = None) -> None:
+        """
+        Pops views from the navigation stack until a view with the given
+        `route` is found, then delivers `result` via the
+        [`on_views_pop_until`][flet.Page.on_views_pop_until] event.
```
Comment (ndonkoHenri) [inline sdk/python/packages/flet/src/flet/controls/page.py]: Update cross refs here too.

File: CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 * Add `ft.use_dialog()` hook for declarative dialog management from within `@ft.component` functions, with frozen-diff reactive updates and automatic open/close lifecycle ([#6335](https://github.com/flet-dev/flet/pull/6335)) by @FeodorFitsner.
 * Add `scrollable`, `pin_leading_to_top`, and `pin_trailing_to_bottom` properties to `NavigationRail` for scrollable content with optional pinned leading/trailing controls ([#1923](https://github.com/flet-dev/flet/issues/1923), [#6356](https://github.com/flet-dev/flet/pull/6356)) by @ndonkoHenri.
+* Add `Page.pop_views_until()` to pop multiple views and return a result to the destination view ([#6326](https://github.com/flet-dev/flet/issues/6326), [#6347](https://github.com/flet-dev/flet/pull/6347)) by [@brunobrown](https://github.com/brunobrown).
```
Comment (ndonkoHenri) [inline CHANGELOG.md]:
```suggestion
* Add `Page.pop_views_until()` to pop multiple views and return a result to the destination view ([#6326](https://github.com/flet-dev/flet/issues/6326), [#6347](https://github.com/flet-dev/flet/pull/6347)) by @brunobrown.
```

**Comment by ndonkoHenri** [comment]
Will it make sense to show the added example in the docs website too? Perhaps on the [nav and routing page](https://flet.dev/docs/cookbook/navigation-and-routing), with a description.

**Comment by brunobrown** [comment]
> Will it make sense to show the added example in the docs website too? Perhaps on the [nav and routing page](https://flet.dev/docs/cookbook/navigation-and-routing), with a description.

Great idea! I'll add a section with an example related to that.

**Comment by FeodorFitsner** [comment]
Great job, @brunobrown!

**Comment by brunobrown** [comment]
Thank you so much, @FeodorFitsner and @ndonkoHenri, for all your guidance and suggestions throughout the contribution process.
